clang/lib/CodeGen/CodeGenFunction.cpp

```diff
@@ -564,6 +564,21 @@ CodeGenFunction::DecodeAddrUsedInPrologue(llvm::Value *F,
                             "decoded_addr");
 }
 
+void CodeGenFunction::EmitSubGroupMetadata(const FunctionDecl *FD,
```

Comment:
It looks like we can re-use `EmitSubGroupMetadata` function to emit `intel_reqd_sub_group_size` in case if we compile OpenCL or SYCL device.

Reply:
There are few issues to do it:

1. We don't generate OpenCL kernel attribute on Host. In case  if we compile OpenCL or SYCL Device, we need to add check like CodeGenFunction::EmitOpenCLKernelMetadata. This will cause conflicts and will not generate IR/AST on Host.

  if (!FD->hasAttr<OpenCLKernelAttr>())
    return;

2.  We still need to emit intel_reqd_sub_group_size inside CodeGenFunction::EmitOpenCLKernelMetadata in case if we compile device/OpenCL, since test/CodeGenSYCL/sycl-multi-kernel-attr.cpp checks multiple attributes together such as intel::reqd_sub_group_size(4) and cl::reqd_work_group_size. 

class Functor {
public:
  [[intel::reqd_sub_group_size(4), cl::reqd_work_group_size(32, 16, 16)]] void operator()() const {}
};

cl::reqd_work_group_size is enabled on device currently, so i decided to generate metadata for function attribute [[intel::reqd_sub_group_size()]] on host and device separately to avoid those situations. 

@Fznamznon and @premanandrao , I am open for any suggestion here. Thanks.

Reply:
Can we just process IntelReqdSubGroupSizeAttr  first in "EmitSubGroupMetadata"? Is there any problem if we have the check for OpenCLKernelAttr after IntelReqdSubGroupSizeAttr  is processed? I think attribute will then be processed for both host and device if we modify the condition wherever EmitSubGroupMetadata is called to call it for host as well. 

I'm not sure if this is a 'good/better' approach though. So I leave it up to you/other reviewers. If we are keeping 2 separate functions to emit metadata on host and device, you can still remove line 630-637 and call EmitSubGroupMetadata instead.

Reply:
I have updated codegen part so that we can re-use EmitSubGroupMetadata function to emit intel_reqd_sub_group_size in case if we compile OpenCL or SYCL device.

Reply:
I'm not sure I understand the changes you made in most recent patch. Why is the additional call to `GenOpenCLArgMetadata ` required?

Reply:
Yes, this is not required. I have fixed this with my current patch. Thanks for catching this.